### PR TITLE
Improve YOLOX and RTMDET performance 

### LINKED ISSUE #6555: YOLOX produces poor results comparing with other models
YOLOX produces poor metric and wrong inference results. Probably due to pre-processing augmentations. Consider reviewing Mosaic, MixUP and add YOLO version of mosaic since it works much better in ultralytics branch. Check mixup as well during this investigation.

**Describe the bug**

**Steps to Reproduce**

1. Train YOLOX on the dataset
2. Run inference, see metric
3.
4.

**Environment:**

- OS: 
- Framework version: 
- Python version:
- OpenVINO version:
- CUDA/cuDNN version:
- GPU model and memory:

### LINKED ISSUE #6567: RTMDET Inst model produces 0% mAP
**Describe the bug**

**Steps to Reproduce**

1. Create a project with simple dataset: horses in my case
2. Annotate few images and run training
3. All MaskRCNN models produces 100% mAP, RTMDER outputs 0%

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

**Comment by leoll2** [comment]
It works, with this PR and #6616 combined, I get excellent predictions with YOLOX on XPU

image

File: library/src/getitune/data/augmentation/transforms.py [hunk 1]
```diff
@@ -21,6 +20,78 @@
...
+    """Create a lightweight cache entry with cloned tensor data.
+
+    Cost: ~3ms for a 3x640x640 float32 image (memcpy only),
+    vs. ~260ms for ``copy.deepcopy`` on a full BaseSample.
+    """
+    masks = getattr(sample, "masks", None)
+    return _CachedSample(
+        image=sample.image.clone(),
+        bboxes=sample.bboxes.clone(),
+        label=sample.label.clone(),
```
Comment (leoll2) [inline library/src/getitune/data/augmentation/transforms.py]: Pyrefly raises these errors here:
> ERROR Object of class `BaseSample` has no attribute `bboxes` [missing-attribute]
> ERROR Object of class `BaseSample` has no attribute `label` [missing-attribute]

The errors make sense because `sample` is type-hinted as `BaseSample`. Is this function supposed to be compatible with all tasks, including e.g. `ClassificationSample`? If not, I suggest to refine the type-hint: `DetectionSample | InstanceSegmentationSample` or similar.

Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py]: I will fix them, the PR is still draft

**Comment by kprokofi** [comment]
image

YOLOX before fix: 0.275 mAP50

**Comment by kprokofi** [comment]
The fix targets mostly RTMDET and YOLOX, for DFINE and DEIMV2 almost no affect considering small dataset. 

image

File: library/src/getitune/recipe/detection/yolox_l.yaml [hunk 2]
```diff
@@ -65,22 +65,20 @@ overrides:
...
             img_scale: $(input_size)
             alpha: 1.5
             p: 0.5
             random_pop: false
             max_cached_images: 10
+        - class_path: getitune.data.augmentation.transforms.Resize
+          init_args:
+            size: $(input_size)
+            keep_aspect_ratio: true
+            pad_value: 114
```
Comment (leoll2) [inline library/src/getitune/recipe/detection/yolox_l.yaml:81]: Is `Resize` applied before or after `Normalize`? If it's before, shouldn't `pad_value` be in range [0-1] instead of 114?

File: application/backend/app/models/training_configuration/augmentation.py [hunk 3]
```diff
@@ -365,6 +365,27 @@ class Mosaic(BaseAugmentationParameter):
...
+        le=1.0,
+        default=0.5,
+        title="Scale",
+        description=(
+            "Scale factor for random perspective crop applied after mosaic assembly. "
+            "The random scale is sampled from [1-scale, 1+scale]. "
+            "Higher values produce more zoom variation."
+        ),
+    )
+    translate: float = Field(
```
Comment (leoll2) [inline application/backend/app/models/training_configuration/augmentation.py:379]: Just a reminder to check, before merging, that these new parameters are correctly displayed in the UI

File: library/src/getitune/recipe/detection/yolox_l.yaml (hunk 2 above)
Comment (leoll2) [inline library/src/getitune/recipe/detection/yolox_l.yaml:81]: Where does the value 114 come from? I understand it's a "gray" value, but it's not exactly in the middle of the range. Why not 127.5 (255/2) instead?

File: library/src/getitune/recipe/detection/yolox_l.yaml [hunk 4]
```diff
@@ -65,22 +65,20 @@ overrides:
...
             img_scale: $(input_size)
-        - class_path: torchvision.transforms.v2.RandomResizedCrop
-          init_args:
-            size: $(input_size)
-            scale: [0.25, 1.0]
-            ratio: [0.75, 1.333]
-            antialias: true
-        - class_path: torchvision.transforms.v2.SanitizeBoundingBoxes
-          init_args:
-            min_size: 1
```
Comment (leoll2) [inline library/src/getitune/recipe/detection/yolox_l.yaml:76]: Why removed? Did it cause any issue?

File: library/src/getitune/backend/openvino/models/instance_segmentation.py [hunk 5]
```diff
@@ -155,7 +155,18 @@ def _customize_outputs(
                 ),
             )
             scores.append(torch.tensor(output.scores.reshape(-1)))
-            masks.append(torch.tensor(output.masks))
+
+            # Rescale masks from model input coords to original image coords.
+            raw_masks = torch.tensor(output.masks)
+            rescaled_masks = rescale_masks_to_original(
```
Comment (leoll2) [inline library/src/getitune/backend/openvino/models/instance_segmentation.py:161]: Can you elaborate on this change? If previously we didn't rescale the masks to the original coordinates, how could the model generate good predictions?

File: library/src/getitune/data/augmentation/pipeline.py [hunk 6]
```diff
@@ -271,6 +273,21 @@ def std(self) -> tuple[float, float, float] | None:
...
+        Delegates to ``pre_cache(dataset)`` on each augmentation that
+        implements this method (e.g. CachedMosaic, CachedMixUp).
+        Call before creating a multi-worker DataLoader so all workers
+        inherit full, diverse caches.
+
+        Args:
+            dataset: The training dataset (must support ``len()`` and ``[]``).
+        """
+        for aug in self.augmentations:
+            if isinstance(aug, (CachedMosaic, CachedMixUp)):
```
Comment (leoll2) [inline library/src/getitune/data/augmentation/pipeline.py]: To avoid hardcoding a bunch of augmentations here, a more generalizable approach  is to define an interface
```python
class CacheableAugmentation(ABC):
    @abstractmethod
    def pre_cache(self, dataset: VisionDataset) -> None:
        ....
```
and make the augmentation classes inherit from it:
```python
class CachedMosaic(tvt_v2.Transform, CacheableAugmentation):
    ....
```
Then this line can be rewritten as:
```python
if isinstance(aug, CacheableAugmentation):
```

File: library/src/getitune/data/augmentation/transforms.py [hunk 7]
```diff
@@ -233,25 +273,33 @@ def forward(self, *inputs: BaseSample) -> BaseSample:  # type: ignore[override]
 
 
 class CachedMosaic(tvt_v2.Transform):
-    """Cached Mosaic augmentation for detection / instance segmentation.
+    """Cached Mosaic augmentation with built-in random perspective crop.
 
-    Combines four images into a 2x mosaic canvas.  The output is
-    the raw 2x canvas.
+    Combines four images into a 2x mosaic canvas, then applies a center-based
```
Comment (leoll2) [inline library/src/getitune/data/augmentation/transforms.py:246]: If I understand correctly, `CachedMosaic` samples 4 images from the training set and stiches them. The cache size is 40 by default. For small datasets, it makes sense to me that such cache avoids frequent reloading of the same images. But when the dataset is large (>1000), the chance of a cache hit seems very small (40/1000 = 4%) for the cache to provide any benefit. What am I missing?

File: library/src/getitune/data/augmentation/transforms.py [hunk 8]
```diff
@@ -451,20 +541,252 @@ def forward(self, *_inputs: BaseSample) -> BaseSample:
...
+    def _random_perspective_crop(
+        self,
+        image: torch.Tensor,
+        bboxes: torch.Tensor,
+        labels: torch.Tensor,
+        masks: torch.Tensor | None,
+        out_h: int,
+        out_w: int,
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor | None]:
+        """Apply center-based random affine crop matching upstream RandomPerspective.
```
Comment (leoll2) [inline library/src/getitune/data/augmentation/transforms.py]: I suggest to add a link to the upstream implementation, for reference

**Comment by leoll2** [comment]
Out of all the changes introduced in this PR for the YOLO models, which one improved the performance the most, according to your experiments?

File: library/src/getitune/recipe/detection/deim_dfine_l.yaml [hunk 9]
```diff
@@ -111,10 +111,6 @@ callbacks:
                 - class_path: torchvision.transforms.v2.SanitizeBoundingBoxes
                   init_args:
                     min_size: 1
-                - class_path: getitune.data.augmentation.transforms.Resize
-                  init_args:
-                    size: [640, 640]
-                    keep_aspect_ratio: false
```
Comment (leoll2) [inline library/src/getitune/recipe/detection/deim_dfine_l.yaml:117]: Why removed? Are DFINE models also in scope of this PR?

File: library/src/getitune/data/augmentation/pipeline.py (hunk 6 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/pipeline.py]: Yeah, we can implement Mixin for these two classes

File: library/src/getitune/backend/openvino/models/instance_segmentation.py (hunk 5 above)
Comment (kprokofi) [reply library/src/getitune/backend/openvino/models/instance_segmentation.py:161]: Seems like it was accidentally removed or never added, I need to investigate. That's why we see mAP 0 or -100% in application for OV models. Predictions are not affected, only metric

File: library/src/getitune/recipe/detection/yolox_l.yaml (hunk 2 above)
Comment (kprokofi) [reply library/src/getitune/recipe/detection/yolox_l.yaml:81]: The value of 114 approximates channel-averaged mean pixel intensity of ImageNet. 
[123.675,116.28,103.53] : 114

This value is commonly used in codebases such as YOLO, made

Comment (kprokofi) [reply library/src/getitune/recipe/detection/yolox_l.yaml:81]: We scale values inside augmentation to match 0:1 range, we can set here from 0:255 value or 0:1, doesn't matter.

**Comment by kprokofi** [comment]
> Out of all the changes introduced in this PR for the YOLO models, which one improved the performance the most, according to your experiments?

The CachedMosaic rewrite — it added proper random perspective/affine transformation after stitching the 4-image mosaic, which was previously missing. Without that geometric augmentation step, the mosaic provided limited diversity regardless of cache state. I found that experimenting with YOLO.
Next: pre-warm cache mechanism provided also improvement: +0.095 val/map_50 (0.3637 → 0.4584) in controlled experiments with identical settings. The reason: with spawn multiprocessing and persistent_workers=True, each of the 8 DataLoader workers starts with an empty cache. Without pre-warming, workers independently build their caches from scratch — meaning early epochs have degraded mosaic quality (drawing companion images from a pool of 1–5 items instead of 20). "Pre-warming" populates the full cache before workers are spawned, so all workers inherit a diverse 20-image pool from epoch 1.

File: library/src/getitune/data/augmentation/transforms.py (hunk 7 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py:246]: The cache is not a traditional lookup cache — it is the source pool for mosaic composition. For each training sample, CachedMosaic needs 3 additional images to stitch a 4-image mosaic. Without a cache, you'd need to randomly load, decode, and transform 3 extra images  on every __getitem__ call. Instead, the cache keeps the N most recently processed samples in memory, and companion images are drawn directly from this pool.

So even with a 1000-image dataset and cache size 40: every mosaic is composed from the 40 cached samples. The cache rotates via FIFO or random eviction, cycling different images through over time. The tradeoff is diversity (larger cache = more varied mosaics) vs memory (each entry holds decoded tensors + annotations). So, we keep a small in-memory pool of recent samples so that each mosaic can draw its 3 companion images.

File: library/src/getitune/recipe/detection/yolox_l.yaml (hunk 4 above)
Comment (kprokofi) [reply library/src/getitune/recipe/detection/yolox_l.yaml:76]: No, we don't need it anymore, Mosaic does it inside now

File: library/src/getitune/recipe/detection/deim_dfine_l.yaml (hunk 9 above)
Comment (kprokofi) [reply library/src/getitune/recipe/detection/deim_dfine_l.yaml:117]: CachedMosaic now converts grid to target resolution, no need for Resize. Since we don't change the DEIM pipeline in application, we can simply remove it. In YOLO recipes I added it since if we remove Mosaic in application, the images should be resized.

File: library/src/getitune/data/augmentation/pipeline.py (hunk 6 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/pipeline.py]: Done https://github.com/open-edge-platform/training_extensions/pull/6614/changes/ce5060d7d8547080cda6a1fa8efba574e1effe95

File: application/backend/tests/unit/api/schemas/test_training_configuration_view.py [hunk 10]
```diff
@@ -730,6 +730,39 @@ def fxt_training_configuration_view_json() -> dict:
...
+                                        "allowed_values": None,
+                                        "depends_on": None,
+                                    },
+                                    {
+                                        "type": "parameter",
+                                        "key": "translate",
+                                        "name": "Translate",
+                                        "description": (
+                                            "Translation fraction for random perspective crop. "
+                                            "The crop center shifts by up to \u00b1translate * image_size pixels."
```
Comment (AlbertvanHouten) [inline application/backend/tests/unit/api/schemas/test_training_configuration_view.py]:
```suggestion
                                            "The crop center shifts by \u00b1translate * image_size pixels."
```

File: library/src/getitune/data/augmentation/transforms.py (hunk 8 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py]: Done

File: library/src/getitune/data/augmentation/transforms.py (hunk 7 above)
Comment (AlbertvanHouten) [reply library/src/getitune/data/augmentation/transforms.py:246]: Wouldn't this still lead to a huge overrepresentation of images that are in the cache being part of the mosaic? The cache rotation only swaps out 1 image every 20 images on average so over 1000 images there will only have been 50 swapped + 20 start images in the cache which make up 3/4th of every mosaic.

File: library/src/getitune/data/augmentation/transforms.py [hunk 11]
```diff
@@ -451,20 +464,245 @@ def forward(self, *_inputs: BaseSample) -> BaseSample:
...
+                masks = mask_crop[valid]
+        else:
+            new_bboxes = bboxes
+
+        return output, new_bboxes, labels, masks
+
+    @typing.no_type_check
+    def forward(self, *_inputs: BaseSample) -> BaseSample:
+        """Apply CachedMosaic with random perspective crop. Output is always img_scale."""
+        assert len(_inputs) == 1, "Only single sample input is supported"  # noqa: S101
```
Comment (AlbertvanHouten) [inline library/src/getitune/data/augmentation/transforms.py]: Please don't use asserts in production code, it should only be used in testing. Python will remove asserts completely if its not run with a debugger which is likely in docker deployments. Just raise an exception.

Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py]: Sure, I will remove it. Sorry, generated shit by AI, I didn't notice that

File: library/src/getitune/data/augmentation/transforms.py (hunk 7 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py:246]: Good point, but pre-warming changes the situation. After `pre_cache` + `freeze_cache`, the cache uses stochastic refresh at a random index. With refresh_rate=0.05 and 1000 forward calls, ~50 replacements happen at uniformly random indices, so ~92% of the pool is refreshed.
Also: the cache is a sampling pool, not duplication. Each mosaic = 1 fresh anchor + 3 cached companions. A cached image is picked per slot with probability 1/20=5%, so even a long-lived entry appears in only ~15% of mosaics.

It still some kind of workaround to solve 2 problems: 1) workers fragmentation due to spawn mp process 2) diversity of the mosaic images. Ideally we should pool from the whole dataset and implement it in collate function or during data loader iterations. Anyway it changes the architecture and will work with a lot of workarounds since we should control it from configs and application

File: application/backend/app/models/training_configuration/augmentation.py (hunk 3 above)
Comment (kprokofi) [reply application/backend/app/models/training_configuration/augmentation.py:379]: image

**Comment by kprokofi** [comment]
image
However, confidence threshold is too high for the model, obviously. After training I see 0.7 in model XML. It filters out most of the predictions. I will check it in another PR

File: library/src/getitune/data/augmentation/transforms.py (hunk 11 above)
Comment (kprokofi) [reply library/src/getitune/data/augmentation/transforms.py]: Done

**Comment by kprokofi** [comment]
image

**Comment by kprokofi** [comment]
image

**Review by leoll2 — APPROVED** [review]

**Review by AlbertvanHouten — APPROVED** [review]
